development/webpack/utils/plugins/SelfInjectPlugin/index.ts

```diff
@@ -0,0 +1,175 @@
+  // It generates the absolute url of the given file as an extension url.
+  // e.g., `chrome-extension://<extension-id>/scripts/inpage.js`
+  sourceUrlExpression: (filename: string) =>
+    `(globalThis.browser||chrome).runtime.getURL(${JSON.stringify(filename)})`,
+} satisfies SelfInjectPluginOptions;
+
+/**
+ * Modifies processed assets to inject a script tag that will execute the asset
+ * as an inline script. Primarily used in Chromium extensions that need to
+ * access a tab's `window` object from a `content_script`.
+ *
+ * @example
+ * Input:
+ * ```js
+ * // webpack.config.js
+ * module.exports = {plugins: [new SelfInjectPlugin({ test: /\.js$/ })]};
+ * ```
+ *
+ * ```js
+ * // src/index.js
+ * console.log("hello world");
+ * ```
+ * Output:
+ * ```js
+ * // dist/main.js
+ * {let d=document,s=d.createElement('script');s.textContent="console.log(\"hello world\");\n//# sourceMappingURL=main.js.map"+`\n//# sourceURL=${(globalThis.browser||chrome).runtime.getURL("main.js")};`;d.documentElement.appendChild(s).remove()}
+ * ```
+ * ```json
+ * // dist/main.js.map (example)
+ * {"version":3,"file":"x","mappings":"AAAAA,QAAQC,IAAI","sources":["webpack://./src/index.js"],"sourcesContent":["console.log(\"hello world\");"],"names":["console","log"]}
+ * ```
+ */
+export class SelfInjectPlugin {
+  private options: SelfInjectPluginOptions & typeof defaultOptions;
+
+  constructor(options: SelfInjectPluginOptions) {
+    validate(schema, options, { name: SelfInjectPlugin.name });
+
+    this.options = { ...defaultOptions, ...options };
+  }
+
+  apply(compiler: Compiler): void {
+    compiler.hooks.compilation.tap(SelfInjectPlugin.name, (compilation) => {
+      this.processAssets(compilation);
+    });
+  }
+
+  /**
+   * Hooks into the compilation process to modify assets.
+   *
+   * @param compilation
+   */
+  processAssets(compilation: Compilation): void {
+    const opts = {
+      name: SelfInjectPlugin.name,
+      stage: Compilation.PROCESS_ASSETS_STAGE_DEV_TOOLING,
+    };
+    compilation.hooks.processAssets.tap(opts, () => this.process(compilation));
+  }
+
+  /**
+   * Processes compilation assets to inject a script tag that will execute the
+   * asset as an inline script.
+   *
+   * @param compilation
+   */
+  process(compilation: Compilation): void {
+    const { test } = this.options;
+    const match = ModuleFilenameHelpers.matchObject.bind(null, { test });
+
+    for (const chunk of compilation.chunks) {
+      for (const file of chunk.files) {
+        if (match(file)) {
+          compilation.updateAsset(file, (asset: Source) => {
+            return this.updateAsset(compilation, file, asset);
+          });
+        }
+      }
+    }
+  }
+
+  /**
+   * Updates the given asset to inject a script tag that will execute the asset
+   * as an inline script.
+   *
+   * @param compilation
+   * @param file
+   * @param asset
+   */
+  updateAsset(compilation: Compilation, file: string, asset: Source): Source {
```

Comment:
This is the mostly same functionality as https://github.com/MetaMask/metamask-extension/blob/63851c9d055d5afa84e04762e615e9ea3cddd805/development/build/utils.js#L265-L270, but the webpack way.

This method has one additional benefit, in that it supports source maps in dev tools for the injected inpage.js content. Pretty neat.